### PR TITLE
General purpose instructions for bytes: Batch 1

### PR DESCRIPTION
**Pull request by kwannoel**
Implements batch 1 features from https://github.com/fare/gerbil-ethereum/issues/49.

Tests pass locally.

### CODE REVIEW

File: t/types-integrationtest.ss [hunk 1]
```diff
@@ -0,0 +1,142 @@
+(export #t)
```
Comment (kwannoel) [inline t/types-integrationtest.ss:1]: Skip this file for now. This is copy pasted from other integration tests. To be modified.

File: types.ss [hunk 2]
```diff
@@ -20,6 +20,7 @@
   :clan/poo/brace
   ./hex ./abi ./rlp)
 
+
```
Comment (kwannoel) [inline types.ss:23]: Remove

File: assembly.ss [hunk 3]
```diff
@@ -423,6 +424,9 @@
       ;; [16B, 38G] -- note that we could skip the ending POP
       (&begin DUP1 n-bytes ADD MLOAD (&shr n-bits) DUP3 (&shl (- 256 n-bits)) OR SWAP1 MSTORE POP)))))
 
+;; n-bytes is a natural in range: [0,32]
+;; addr is the memory offset
+;; Stack: | value | ... |
```
Comment (kwannoel) [inline assembly.ss:429]: remove

File: assembly.ss [hunk 4]
```diff
@@ -535,3 +536,68 @@
...
+;; ----------------------------------
+;; Utilities for variable-size values
+;; ----------------------------------
+;; Storage Convention:
+;; Values are stored in big endian form.
+;; Values larger than an EVM Word are contiguously stored, while preserving byte-order.
+;;
+;; For Example, suppose we have the following value:
+;;
+;; 0x ff ff ff    ...    ff aa 11
```
Comment (fare) [inline assembly.ss:549]: Unclear what this is being represented. A 33-byte integer? Some record with a 31-byte integer field, followed by two 1-byte fields?

File: assembly.ss [hunk 5]
```diff
@@ -2,7 +2,8 @@
 
 (import
   :gerbil/gambit/bits :gerbil/gambit/bytes :gerbil/gambit/exact
-  :std/misc/bytes :std/misc/number :std/sugar
+  :std/format :std/misc/bytes :std/misc/list :std/misc/number
```
Comment (kwannoel) [inline assembly.ss:5]: Remove import

File: evm-runtime.ss [hunk 6]
```diff
@@ -661,3 +661,74 @@
...
+;; Bytes Stack/memory representation follows byte-order (big-endian).
+;;
+;; For example, given the value: 0x00ff,
+;; After PUSH:  Stack will have "0x00" on top, followed by "0xff".
+;; After STORE: Memory location will have "0x00" at offset, "0xff" at offset+1.
+
+;; stack output: value/0 value/1 ... value/n
+;; Pushes bytes onto the EVM,
+;; in 32-byte chunks and the remaining bytes (if any).
+;; We have to PUSH later bytes first,
```
Comment (kwannoel) [inline evm-runtime.ss:679]: Add NOTE:

File: evm-runtime.ss [hunk 7]
```diff
@@ -661,3 +661,74 @@
...
+(def &mload-1/any-size
+  (match <>
+    ([relative-offset size]
+     (&begin
+      DUP1 relative-offset ADD
+      (&mload size) SWAP1))))
+
+;; stack input:  start-offset
+;; stack output: bytes/0 bytes/1 ... bytes/n
+;; TODO test when size = 0
```
Comment (kwannoel) [inline evm-runtime.ss:723]: Remove

File: t/80-evm-eval-integrationtest.ss [hunk 8]
```diff
@@ -263,4 +263,141 @@
...
+
+      (evm-test [] (&begin
+                    #x20 (&mstoreat brk@ 32)
+                    &push-true)
+                [[UInt8 . 1]]
+                result-in-memory?: #t
+                result-start: #x20))
+
+    (test-case "EVM-type/&mem<-stack<- size = 65"
+      ;; TODO Integrate these with existing EVM types,
```
Comment (kwannoel) [inline t/80-evm-eval-integrationtest.ss:317]: Remove

File: assembly.ss (hunk 4 above)
Comment (kwannoel) [reply assembly.ss]: Updated here: https://github.com/fare/gerbil-ethereum/pull/46/files#diff-61b54f933d2406971aa705cb01aaeb6d4edae919accf5f1009c04e3c082a8c40R672

File: t/80-evm-eval-integrationtest.ss [hunk 9]
```diff
@@ -263,4 +263,163 @@
...
+                    (&push/any-size str65/bytes) ; BAA..A CAA..A D
+                    (&mstore/any-size 65))       ; -
+                [[UInt256 . str65/1/nat]
+                 [UInt256 . str65/2/nat]
+                 [UInt8   . str65/3/nat]
+                 ]
+                result-in-memory?: #t
+                result-start: #x20))
+
+    (test-case "EVM-type/&mem<-stack<- size = 1"
```
Comment (kwannoel) [inline t/80-evm-eval-integrationtest.ss:308]: Move these out into their own integration tests.

File: t/type-descriptors-test.ss [hunk 10]
```diff
@@ -0,0 +1,61 @@
+(export #t)
+
+(import
+  :gerbil/gambit/bytes
```
Comment (kwannoel) [inline t/type-descriptors-test.ss:4]: Purge unnecessary imports

File: evm-runtime.ss [hunk 11]
```diff
@@ -661,3 +661,94 @@
...
+;; stack output: -
+(def (&mstore/any-size size)
+  (assert! (<= 1 size) (format "size: ~d not in range: [1,∞]")) ; TODO: provide a concrete upper bound
+  (def sizes/base-32 (sizes/word-size<-size size))
+  (&begin (map &brk-cons sizes/base-32) ...)
+  )
+
+;; Given byte length,
+;; generates relative offsets and sizes for parts
+;; E.g. (offsets-and-sizes<-size 65) -> [[0 . 32] [32 . 32] [64 . 1]]
```
Comment (kwannoel) [inline evm-runtime.ss:720]: Not pair, list of lists.

File: evm-runtime.ss [hunk 12]
```diff
@@ -661,3 +661,94 @@
...
+;; Pushes bytes onto the EVM,
+;; in 32-byte chunks and remaining bytes (if any).
+;;
+;; NOTE: We have to PUSH earlier bytes last, to preserve byte-order.
+;; E.g. Given 65 bytes:
+;; PUSH bytes[64-65], PUSH bytes[32-64], PUSH bytes[0-32]
+;; stack: bytes[0-32] bytes[32-64] bytes[64-65] ...
+;;
+;; stack input: -
+;; stack output: value/0 value/1 ... value/n
```
Comment (kwannoel) [inline evm-runtime.ss:690]: bytes/0 bytes/1 ...

File: type-descriptors.ss [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+;; For types larger than an EVM Word / variably sized, we classify as boxed.
+;;
+;; --------------
+;; Representation
+;; --------------
+;; First, convert types to their bytes representation.
+;; byte size is used to determine how to store on / load from stack or mem.
+;; Store them to stack or memory accordingly.
+
+;; TODO Separate types used for existing ABI and EVM types??
```
Comment (kwannoel) [inline type-descriptors.ss:47]: Tracked in issue under integrating with EVM types, can remove.

File: type-descriptors.ss [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+  ;; NOTE: These use free memory pointer
+  .&mem<-stack: (cut .&with-value/size <> &mstore/any-size)
+  .&stack<-mem: (cut .&with-value/size <> &mload/any-size)
+  .&mem<-stack<-: (lambda (value) (&begin (.&stack<- value) (.&mem<-stack value)))
+  ;; .&memcpy: ; TODO copy memory from one location to another
+  )
+
+;; TODO pull and read from CALLDATA. maybe from mem -> stack ? stack
+;; TODO right to left arrows naming conventions
```
Comment (kwannoel) [inline type-descriptors.ss:74]: resolved

File: type-descriptors.ss [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+;; TODO abstract stack->mem etc..
```
Comment (kwannoel) [inline type-descriptors.ss:75]: resolved

**Comment by kwannoel** [comment]
Rebased and cleaned up commits.

File: type-descriptors.ss [hunk 16]
```diff
@@ -0,0 +1,72 @@
+;; We are shadowing existing types. Should we monkey-patch them instead? Let's hope not.
```
Comment (kwannoel) [inline type-descriptors.ss:1]: remove copy pasted stuff

File: type-descriptors.ss [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+(export #t)
+(export (import: :clan/poo/mop) (import: :clan/poo/type) (import: :clan/poo/number))
+
+;; TODO cleanup imports
+(import
+  (for-syntax :gerbil/gambit/exact :std/iter :std/stxutil :clan/syntax)
```
Comment (kwannoel) [inline type-descriptors.ss:7]: cleanup imports

File: evm-instructions.ss [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+      DUP1 relative-offset ADD ; offset       start-offset
+      (&mload size) SWAP1))))  ; start-offset bytes[offset:end]
+
+;; stack input:  start-offset
+;; stack output: part0 part1 ... partn
+(def (&mload/any-size size)
+  (assert! (<= 0 size) (format "size: ~d not in range: [1,∞]")) ; TODO: provide a concrete upper bound
+  (match (offsets-and-sizes<-size size)
+    ;; = 0 bytes
+    ([] (&begin POP 0)
```
[the remark this answers is not in the document]
Comment (fare) [reply evm-instructions.ss:101]: Either unit types should be statically optimized away, or should dynamically have some marker, but whatever the convention is, we must make it consistent.

File: t/100-evm-instructions-integrationtest.ss [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+                            DUP1 64 ADD       ; -- offset+64   offset
+                            (&mload 1) SWAP1  ; -- offset      bytes[64-65]
+
+                            ;; Load second string segment
+                            DUP1 32 ADD       ; -- offset+32   offset       bytes[64-65]
+                            (&mload 32) SWAP1 ; -- offset      bytes[32-64] bytes[64-65]
+
+                            ;; Load first string segment
+                            (&mload 32)       ; -- bytes[0-32] bytes[32-64] bytes[64-65]
+                            )))
```
Comment (fare) [inline t/100-evm-instructions-integrationtest.ss:107]: parens

File: t/80-evm-eval-integrationtest.ss [hunk 20]
```diff
@@ -263,4 +263,16 @@
...
+    (test-case "&mstore 1 byte"
+      (evm-test [] (&begin 42 0 (&mstore 1))
+                [[UInt8 . 42]]
+                result-in-memory?: #t))
+
+    (test-case "&mstore 32 bytes"
+      (def maxUInt256 (- (expt 2 256) 1))
+      (evm-test [] (&begin maxUInt256 0 (&mstore 32))
+                [[UInt256 . maxUInt256]]
+                result-in-memory?: #t))
```
Comment (fare) [inline t/80-evm-eval-integrationtest.ss:277]: parens

File: t/evm-instructions-test.ss [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+      (def 65-bytes (list->u8vector (make-list 65 65)))
+      (u8vector-set! 65-bytes 0 66)  ; bytes/0-32:  66 65 65 ... 65
+      (u8vector-set! 65-bytes 32 67) ; bytes/32-64: 67 65 65 ... 65
+      (u8vector-set! 65-bytes 64 68) ; bytes/64-65: 68
+
+      (check-inst? (&push/any-size 65-bytes)
+                   [(subu8vector 65-bytes 64 65) ; bytes/64-65: 68
+                    (subu8vector 65-bytes 32 64) ; bytes/32-64: 67 65 65 ... 65
+                    (subu8vector 65-bytes 0  32) ; bytes/0-32:  66 65 65 ... 65
+                    ])
```
Comment (fare) [inline t/evm-instructions-test.ss:37]: parens

**Review by fare — APPROVED** [review]

File: evm-instructions.ss [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+  (def words (make-list n-words word-size))
+  (def rem (modulo size word-size))
+  (if (eq? rem 0) words [words ... rem])
+  )
+
+;; NOTE: Uses brk@ for offset via &brk-cons, dependent on EVM memory layout.
+;; stack input:  part0 part1 ... partn
+;; stack output: -
+(def (&mstore/free/any-size size)
+  (assert! (<= 1 size) (format "size: ~d not in range: [1,∞]")) ; TODO: provide a concrete upper bound
```
Comment (zenhack) [inline evm-instructions.ss:66]: Whatever we decide to do re: validating the size of inputs, we should probably factor it out into a shared function; you have the same assert (with the same TODO) elsewhere.

File: evm-instructions.ss [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+;; - Storing to memory with reference
+;; - Load from memory with reference
+;; NOTE: Reference is start-offset, length on stack.
+;;
+;; -------------
+;; Byte encoding
+;; -------------
+;; Instructions encode bytes as big-endian.
+;;
+;; For example, given the value: 0x00ff,
```
[the remark this answers is not in the document]
Comment (zenhack) [reply evm-instructions.ss:25]: I agree this is confusing; I would expect this to live in one slot on the stack, since it fits in 2 bytes and the word size is 32 bytes.

File: evm-instructions.ss [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+
+;; Helper function - Make list of relative offsets and sizes for partitions.
+;; E.g. (offsets-and-sizes<-size 65) -> [[0 32] [32 32] [64 1]]
+;; List [RelativeOffset Size] <- Nat
+(def (offsets-and-sizes<-size size)
+  (def sizes (sizes/word-size<-size size))
+  (def relative-offsets (iota (length sizes) 0 32))
+  (zip relative-offsets sizes))
+
+;; Given relative-offset and size,
```
Comment (zenhack) [inline evm-instructions.ss:79]: Should specify a formal signature here. It isn't immediately obvious how the arguments are passed in.

File: evm-instructions.ss (hunk 18 above)
[the remark this answers is not in the document]
Comment (zenhack) [reply evm-instructions.ss:101]: Agree; my feeling is that `(&mload/any-size 0)` should not push anything onto the stack.

**Review by zenhack — CHANGES_REQUESTED** [review]
Left some comments in-line

File: evm-instructions.ss [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+
+;; ----------------------
+;; Instruction defintions
+;; ----------------------
+
+;; stack input: -
+;; stack output: part0 part1 ... partn
+(def (&push/any-size bytes (start 0) (len (u8vector-length bytes)))
+  (assert! (<= 1 len) (format "size: ~d not in range: [1,∞]")) ; TODO: provide a concrete upper bound
+  (def end (min len (+ start 32)))
```
[the remark this answers is not in the document]
Comment (fare) [reply evm-instructions.ss:45]: We do care about supporting data structures sizes that are not multiple of 32, because storage is so expensive that if we can save one word of SSTORE it will save 5000 GAS which at current prices is about 15 cents. Per store, per invocation. Even one saved reading would be 2100 GAS, or over 6 cents.

I believe the current `&push` assumes left-padding. We could keep that. Or we could support the Solidity ABI of left-padding for numbers, right-padding for bytes. They are the two options that make sense. Always left-pad seems cheaper both in human time and in EVM runtime.

File: evm-instructions.ss [hunk 26]
```diff
@@ -0,0 +1,115 @@
...
+    )))
+
+;; Helper function - Make list of word-sizes
+;; E.g. (sizes/word-size<-size 65 32) -> [32 32 1]
+(def (sizes/word-size<-size size (word-size 32))
+  (def n-words (truncate (/ size word-size)))
+  (def words (make-list n-words word-size))
+  (def rem (modulo size word-size))
+  (if (eq? rem 0) words [words ... rem])
+  )
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss]: Thanks for pointing this out.

File: assembly.ss (hunk 4 above)
Comment (kwannoel) [reply assembly.ss]: Follow up thread: https://github.com/fare/gerbil-ethereum/pull/46#discussion_r667566744

File: evm-instructions.ss [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+    (let ()
+      (def bytes<=32 (subu8vector bytes start end))
+      (def n-bytes (u8vector-length bytes<=32))
+      (&begin (&push/any-size bytes end) [&push-bytes bytes<=32])
+    )))
+
+;; Helper function - Make list of word-sizes
+;; E.g. (sizes/word-size<-size 65 32) -> [32 32 1]
+(def (sizes/word-size<-size size (word-size 32))
+  (def n-words (truncate (/ size word-size)))
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss:55]: Thanks for these.

Where can I find `floor/` / `floor-quotient` in documentation?
It does not appear in usual references:
- https://schemers.org/Documents/Standards/R5RS/HTML/
- https://www.iro.umontreal.ca/~gambit/doc/gambit.html#General-index
- https://srfi.schemers.org/?q=floor

[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss:55]: Found it in [r7rs](https://small.r7rs.org/attachment/r7rs.pdf).

File: evm-instructions.ss [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+;;         end    = offset+size
+(def &mload-1/any-size
+  (match <>
+    ([relative-offset size]
+     (&begin                   ; start-offset
+      DUP1 relative-offset ADD ; offset       start-offset
+      (&mload size) SWAP1))))  ; start-offset bytes[offset:end]
+
+;; stack input:  start-offset
+;; stack output: part0 part1 ... partn
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss:94]: To clarify: 
`Thunk` refers to code, which when evaluated, will return the expected value, i.e. `part0 ... partn`.

File: evm-instructions.ss (hunk 24 above)
Comment (kwannoel) [reply evm-instructions.ss:77]: Included here: https://github.com/fare/gerbil-ethereum/pull/46/files#diff-6c56dd17a26fa6ddfd679ad2e9545670edacc0322c7e5f2a94199a7385c20fabR86

File: evm-instructions.ss (hunk 18 above)
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss:101]: Was following convention from our `assembly.ss` code:
```
(def (&mload n-bytes)
  (assert! (and (exact-integer? n-bytes) (<= 0 n-bytes 32)) "Bad length for &mload")
  (cond
   ((zero? n-bytes) (&begin POP 0)) ;; [3B, 5G]
  ...
```

[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss:101]: Will follow @zenhack 's suggestion and just `POP` the `offset`.

File: evm-instructions.ss (hunk 28 above)
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss:95]: Updated with signature.

File: evm-instructions.ss (hunk 23 above)
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss]: I think my wording was ambiguous.
Instead of the first slot on EVM stack as `0x00` and second slot as `0xff`,
I meant the **first byte** on EVM stack should be `0x00` and **second byte** should be `0xff`.
Each element of the array will not occupy an entire EVM slot on its own.

Have reworded the example.

File: evm-instructions.ss [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+
+
+;; ----------------------
+;; Instruction defintions
+;; ----------------------
+
+;; stack input: -
+;; stack output: part0 part1 ... partn
+(def (&push/any-size bytes (start 0) (len (u8vector-length bytes)))
+  (assert! (<= 1 len) (format "size: ~d not in range: [1,∞]")) ; TODO: provide a concrete upper bound
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss]: > We could also assert that start + len <= length(bytes), rather than just no-oping. Generally better to err on the side of catching mistakes sooner rather than later. This also conveniently obviates picking a specific upper bound.

`start` is updated to point to the start of the section of the array we have not generated thunks for.
`len` is a constant which stores the total number of bytes, rather than the length of the current section.

For instance given:
```
start = 0, length(bytes) = 64
```

During the first iteration, we will load the first 32 bytes i.e. 0...31(inclusive). 

In the second iteration we will have:
```
start = 32, len = 64 ;; See here that len is unchanged.
```

In that case `start + len = 96 > len = 64`.

I have renamed `len` to `total-bytes` to make this distinction clear.
Added in assertions to provide similar guarantees.
```
  (assert! (< start total-bytes))
  (assert! (<= end total-bytes))
  (when (< start end)
    ...
```

File: evm-instructions.ss [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+;; We use the notation: "part[n]", to denote the partition index.
+;; E.g. 65-bytes -> | 32-bytes | 32-bytes | 1-byte  |
+;;                  | part0  | part1  | part2 |
+
+
+;; ----------------------
+;; Instruction defintions
+;; ----------------------
+
+;; stack input: -
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss:44]: `&push/any-size` uses the existing `&push-bytes` instruction provided in `assembly.ss`.
Hence this follows our current left-pad convention.

[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss:53]: > Re: left vs. right padding, another alternative would be to just throw an error if the input is not a multiple of the word size, forcing the caller to make the decision. I don't have a strong opinion here.
> 
> This could also use some comments describing the rest of the contract for the (scheme) function as well as just a high level "what is this for," rather than just the stack signature of the generated code.

Updated.

File: evm-instructions.ss (hunk 25 above)
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss]: > Let's define the magic number 32 as a constant somewhere.

Updated.

[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss]: > We do care about supporting data structures sizes that are not multiple of 32, because storage is so expensive that if we can save one word of SSTORE it will save 5000 GAS which at current prices is about 15 cents. Per store, per invocation. Even one saved reading would be 2100 GAS, or over 6 cents.
> 
> I believe the current `&push` assumes left-padding. We could keep that. Or we could support the Solidity ABI of left-padding for numbers, right-padding for bytes. They are the two options that make sense. Always left-pad seems cheaper both in human time and in EVM runtime.

Yes that is correct. Approach taken is to left-pad, via `&push` /  `&push-bytes`.

File: evm-instructions.ss [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+;; - Storing to memory with reference
+;; - Load from memory with reference
+;; NOTE: Reference is start-offset, length on stack.
+;;
+;; -------------
+;; Byte encoding
+;; -------------
+;; Instructions encode bytes as big-endian.
+;;
+;; For example, given an array of two bytes: 0x00ff:
```
Comment (zenhack) [inline evm-instructions.ss:26]: This is still confusing to me. Things that would I think make this clearer:

- Using `0x00ff` as notation for an array of bytes is weird, since it looks like a single value (which leaves open the question of endianness. You state big-endian above, but it makes it hard to follow).; I would maybe suggest using gerbil notation for sequences of bytes, i.e. `#u8(#x00 #xff)`, to make this less ambiguous.
- Talking about "bytes" on the stack is confusing; the operand stack does not contain bytes, it contains 32-byte values, whose endianness is unspecified. Instead of describing the operand stack as if it contained bytes, I would specify what the value on the top of the stack is after this push -- which if I'm reading this correctly would be `0x00ff000000000000000000000000000000000000000000000000000000000000`

File: evm-instructions.ss (hunk 30 above)
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss:56]: From discussion with @zenhack on discord:
> The stack is still defined in 32-byte chunks, so that leaves open the question of what the other 31 bytes in that last word are after the load. Left padding means you have a bunch of zeros in the most-significant bytes, right padding means you have them in the least significant bytes. It's not possible to have a non-integer number of words on the stack. I agree padding is unnecessary and undesirable in memory.

I have added tests to verify left-padding occurs.

File: evm-instructions.ss (hunk 31 above)
Comment (kwannoel) [reply evm-instructions.ss]: I think I was confused regarding padding.
It should be left padded, so we should expect this instead:
```
#x00000000000000000000000000000000000000000000000000000000000011ff
```

I have added testcases in `t/100-evm-instructions-integrationtest.ss` to give a concrete representation of what is expected:
```
    (test-case "EVM-type: &mstore/free/any-size, size = 1, left-padded to EVM-Word size"
      (evm-test [] (&begin
                    &init-brk
                    (&push/any-size #u8(1))    ; stack: 1
                    (&mstore/free/any-size 32)
                    &store-brk
                    )
                [[Bytes32 . #u8(0 0 0 0  0 0 0 0  0 0 0 0  0 0 0 0
                                0 0 0 0  0 0 0 0  0 0 0 0  0 0 0 1
                                )]
                 [UInt256 . 64]]
                result-in-memory?: #t
                result-start: #x20))

    (test-case "EVM-type: &mstore/free/any-size, size = 4, left-padded to EVM-Word size"
      (evm-test [] (&begin
                    &init-brk
                    (&push/any-size #u8(1 2 3 4)) ; stack: 1 2 3 4
                    (&mstore/free/any-size 32)
                    &store-brk
                    )
                [[Bytes32 . #u8(0 0 0 0  0 0 0 0  0 0 0 0  0 0 0 0
                                0 0 0 0  0 0 0 0  0 0 0 0  1 2 3 4
                                )]
                 [UInt256 . 64]]
                result-in-memory?: #t
                result-start: #x20))
```

Comment (kwannoel) [reply evm-instructions.ss]: Updated the examples.

File: t/100-evm-instructions-integrationtest.ss [hunk 32]
```diff
@@ -0,0 +1,137 @@
...
+                    &init-brk
+                    (&push/any-size #u8(1))   ; stack: 1
+                    (&mstore/free/any-size 1)
+                    &store-brk
+                    )
+                [[Bytes1 . #u8(1)]
+                 [UInt256 . 33]]
+                result-in-memory?: #t
+                result-start: #x20))
+
```
Comment (kwannoel) [inline t/100-evm-instructions-integrationtest.ss:30]: See this section for padding tests @zenhack.

**Review by zenhack — APPROVED** [review]
LGTM

File: evm-instructions.ss (hunk 25 above)
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss]: By default `PUSH` instructions left-pad to evm-word length.

File: evm-instructions.ss (hunk 23 above)
[the remark this answers is not in the document]
Comment (kwannoel) [reply evm-instructions.ss]: See comment: https://github.com/fare/gerbil-ethereum/pull/46#discussion_r673112186, 
for follow-up discussion.

**Comment by kwannoel** [comment]
Rebasing and cleaning up commits (no changes to code).
